### PR TITLE
Merge ign-common4 ➡️  gz-common5

### PR DESCRIPTION
**Pull request by scpeters**
# ➡️  Forward port

Port `ign-common4 ` ➡️  `gz-common5`

Branch comparison: https://github.com/gazebosim/gz-common/compare/gz-common5...ign-common4

**Note to maintainers**: Remember to **Merge** with commit (not squash-merge or rebase)

### CODE REVIEW

**Comment by mjcarroll** [comment]
There are some things failing here that I don't think were failing before.

File: graphics/src/Image_TEST.cc [hunk 1]
```diff
@@ -93,6 +140,44 @@ TEST_F(ImageTest, ImageProperties)
       std::string::npos);
 }
 
+TEST_F(ImageTest, ImageSavePng)
+{
+  std::string pathOut;
+  ASSERT_TRUE(common::testing::TestTmpPath(pathOut));
```
Comment (scpeters) [inline graphics/src/Image_TEST.cc:146]: @mjcarroll `TestTmpPath` now has different implementations for Bazel and CMake in `gz-common5`. In order to forward-port this, should we split this test out into separate files for Bazel and CMake?

Comment (mjcarroll) [reply graphics/src/Image_TEST.cc:146]: Change to use this method: https://github.com/gazebosim/gz-common/blob/7b0a17cda12a6c0916787714995a4a1f094aeeaa/testing/include/gz/common/testing/TestPaths.hh#L184-L202

Which will let you get a file path in a temporary directory for testing, and automatically do the right thing (TM) between cmake and bazel.

For example `auto pathOut = gz::common::testing::TempPath("foo.png")`

Comment (scpeters) [reply graphics/src/Image_TEST.cc]: thanks!

**Comment by scpeters** [comment]
CI is finally happy

**Review by mjcarroll — APPROVED** [review]

**Pull request merged**